source/common/router/config_impl.cc

```diff
@@ -703,6 +732,28 @@ RetryPolicyImpl RouteEntryImplBase::buildRetryPolicy(
   return RetryPolicyImpl();
 }
 
+InternalRedirectPolicyImpl RouteEntryImplBase::buildInternalRedirectPolicy(
+    const envoy::config::route::v3::RouteAction& route_config,
+    ProtobufMessage::ValidationVisitor& validator, absl::string_view current_route_name) const {
+  if (route_config.has_internal_redirect_policy()) {
+    return InternalRedirectPolicyImpl(route_config.internal_redirect_policy(), validator,
+                                      current_route_name);
+  }
+  envoy::config::route::v3::InternalRedirectPolicy policy_config;
+  switch (route_config.internal_redirect_action()) {
```

Comment:
Wait, aren't you deprecating this?
@htuch this is where I get troubled by code cruft.  If we're going to support old and new versions for 2+ years we're going to have cruft for 2 years.
Can we have an ingress translation library which says "if I see an internal redirect action I will convert into an internal redirect policy" so the actual Envoy code only deals with the new?
cc @mattklein123

Reply:
I don't know how we could automatically generate code to transform the old fields to the new thing?

I think though per our community meeting discussion we don't have to live with this for 2+ years if we allow deprecations to continue within some minor version API update. IMO we should discuss this separately as the type of change we are seeing here we have been doing for a long time.

Reply:
We can only automatically translate some very basic things, like renames or package moves.

The PoR that I thought we all agreed to is not to have cruft for 2+ years, it's to start using minor versions and client feature capabilities so we can turn down Envoy implementation support after 1 year.

Reply:
+1 to discussing this separately, insofar as this PR is clearly not going to be blocked on sorting out our N year API policy, but I think this is a great example we can use to come to consensus to bring to the community for wider approval. - will break out this email thread so it doesn't get in the way of the review :-)